src/main/java/com/iexec/core/replicate/ReplicateSupplyService.java

```diff
@@ -118,39 +126,33 @@ public ReplicateSupplyService(ReplicatesService replicatesService,
             contributionTimeoutTaskDetector.detect();
         }
 
-        // TODO : Remove this, the optional can never be empty
-        // This is covered in workerService.canAcceptMoreWorks
-        Optional<Worker> optional = workerService.getWorker(walletAddress);
-        if (optional.isEmpty()) {
-            return Optional.empty();
-        }
-        Worker worker = optional.get();
-
         for (Task task : validTasks) {
             String chainTaskId = task.getChainTaskId();
+            final Optional<ReplicatesList> oReplicatesList = replicatesService.getReplicatesList(chainTaskId);
+            // Check is only here to prevent
+            // "`Optional.get()` without `isPresent()` warning".
+            // This case should not happen.
+            if (oReplicatesList.isEmpty()) {
+                continue;
+            }
+
+            final ReplicatesList replicatesList = oReplicatesList.get();
 
             // no need to go further if the consensus is already reached on-chain
             // the task should be updated since the consensus is reached but it is still in RUNNING status
-            if (taskUpdateManager.isConsensusReached(task)) {
+            if (taskUpdateManager.isConsensusReached(replicatesList)) {
                 taskUpdateManager.publishUpdateTaskRequest(chainTaskId);
                 continue;
             }
 
-            // skip the task if it needs TEE and the worker doesn't support it
-            // TODO : access worker TEE status through workerService to avoid workerService.getWorker call
-            boolean isTeeTask = task.isTeeTask();
-            if (isTeeTask && !worker.isTeeEnabled()) {
-                continue;
-            }
-
             taskService.initializeTaskAccessForNewReplicateLock(chainTaskId);
             if (taskService.isTaskBeingAccessedForNewReplicate(chainTaskId)) {
                 continue;//skip task if being accessed
             }
             taskService.lockTaskAccessForNewReplicate(chainTaskId);
 
             boolean hasWorkerAlreadyParticipated = replicatesService.hasWorkerAlreadyParticipated(
```

Comment:
Getting very nice!

I feel we can merge in a single `if` following items:
* isTaskBeingAccessedForNewReplicate
* isInContributionPhase
* hasWorkerAlreadyParticipated
* taskNeedsMoreContributions
* enclaveChallenge.isEmpty()

Reply:
I'm not sure it'd be great. If we do as you suggest, either we'll have to pre-compute everything so that it'll represent some load, either we'll have to have a long `if` clause. I don't really like any of these cases. As that could be a critical part of our workflow, it should probably be the fastest possible.
I think the best thing to do there is to extract all these checks into another method (`boolean canWorkerAcceptTask(Task task, String walletAddress)`?).

Reply:
So, I've tried something in https://github.com/iExecBlockchainComputing/iexec-core/pull/490/commits/2468dc1dde1b5dba83de7dbb546b5876aaf6e7f8. That's a first step toward [your suggestion about refactoring `getAuthOfAvailableReplicate`](https://github.com/iExecBlockchainComputing/iexec-core/pull/490#issuecomment-1017601472).
Let me know what you think about it!

Reply:
Nice!